### PR TITLE
Rework translate API to allow associated types to be translated

### PR DESCRIPTION
**Pull request by sanket1729**
The Translate API currently does not support translating the hashes and timelocks. Ideally, we want to write an HTLC as `or( and( alice_key, older(timeout), and(bob_key, sha256(hash)))`. 
But it is impossible to use placeholders for `hash`, and `timelock`. It is only possible for keys. 

This is a significant API change, but I think it is for good. The diff is large, but reviewing `lib.rs` is the main diff, other diffs are just chasing compiler errors

### CODE REVIEW

**Comment by sanket1729** [comment]
@tcharding, you may be interested in the latest commit. `rustfmt` on nightly suggested new changes all of a sudden. Both the tips HEAD and HEAD~1 from "Misc rustfmt check"  2e9bd32 pass the `cargo +nightly fmt --all -- --check`

**Review by apoelstra — DISMISSED** [review]
utACK 2e9bd32581d4ed9d418965c22ab556ab458221ab

**Comment by apoelstra** [comment]
As discussed in another channel, I am happy with this approach. I agree the `Translator` trait might involve some boilerplate for a lot of users, but the `PkTranslate` etc ones help with this, and anyway the existing API also involves boilerplate, often with crazy inscrutable syntax like `<_, _, hash160::Hash, _>` and with names like `translate_pk1` `translate_pk2` etc.

**Review by tcharding — COMMENTED** [review]
Nice work bro! I love the simplification of all the translate traits. I had a good long play with this. I pushed 5 patches to a branch on my fork with the same name `reword_translate`. Feel free to use them, ignore them, or steal whatever parts you like. 4 of them are trivial, docs and things. In one patch I remove the `PkTranslator` trait, includes full justification in the commit log.

Cheers

**Comment by sanket1729** [comment]
@tcharding, I looked at the branch. I agree with the four commits modulo removing the Sha256 one. Shortly, we will also have associated types for `Sha256`, `Hash256`, `Ripemd160`, `Hash160`, `Older`, `After`. 

The purpose of the macros is to avoid 20-line pre-conditions on each function.

**Comment by tcharding** [comment]
No sweat, cheers.

**Comment by sanket1729** [comment]
Thanks a lot for the suggestions and improvements. Will address them shortly.

**Comment by sanket1729** [comment]
Addressed all suggestions and cherry-picked the 4 commits as mentioned above.

**Review by tcharding — COMMENTED** [review]
I like the result of these macros, not having the trait bounds everywhere, but they are not pretty. I had a play around and could not come up with anything better though. Someone with better macro skills than me can probably get rid of the trailing comma requirements. A couple of meta comments:
- The changes in commit `f483089 Fix some macro bugs for 1.41.1` are not specific to Rust 1.41.1 as the title suggests, probably should be squashed.
- Commit `2ffa5ea Rename Sha256Hash type to Sha256` could be squashed
- All the commits I did can be squashed into the original work, I don't need attribution
- The rustfmt change would be cleaner at the front of the PR because its totally unrelated

So I'd probably expect this PR to be three patches
- The rustfmt one
- Introduce impl macros
- Introduce `Translator` trait

File: src/macros.rs [hunk 1]
```diff
@@ -16,6 +16,88 @@ macro_rules! policy_str {
...
+
+                $(#[$meta])*
+                fn $fn($($arg: $type)* ) -> $ret {
+                    $body
+                }
+            }
+    };
+}
+
+/// Macro for implementing FromTree trait. This avoids copying all the Pk::Associated type bounds
```
Comment (tcharding) [inline src/macros.rs:75]: This comment has cut'n'paste glitch - `FromTree`

File: src/descriptor/tr.rs [hunk 2]
```diff
@@ -388,10 +388,11 @@ where
     }
 }
 
+#[rustfmt::skip]
 impl_block_str!(
     Tr<Pk>,
     // Helper function to parse taproot script path
-    fn parse_tr_script_spend(tree: &expression::Tree) -> Result<TapTree<Pk>, Error> {
+    fn parse_tr_script_spend(tree: &expression::Tree,) -> Result<TapTree<Pk>, Error> {
```
[the remark this answers is not in the document]
Comment (apoelstra) [reply src/descriptor/tr.rs:391]: I also tried and couldn't do it.

[the remark this answers is not in the document]
Comment (sanket1729) [reply src/descriptor/tr.rs:391]: Should be possible. Will work on it

**Review by apoelstra — DISMISSED** [review]
ACK a086861c1ca28f0442269690690ab78db2505a77

**Comment by sanket1729** [comment]
Squashed and force pushed. Only one line diff fixing the nit

```
sanket1729:~/rust-miniscript$ git diff 8e5b09ba3b47b99d9e44e389a3b81adcb5580aed a086861
diff --git a/src/macros.rs b/src/macros.rs
index 7f59aa9..066beae 100644
--- a/src/macros.rs
+++ b/src/macros.rs
@@ -72,7 +72,7 @@ macro_rules! impl_from_str {
     };
 }
 
-/// Macro for impl Struct with associated bounds. This avoids copying all the Pk::Associated type bounds
+/// Macro for implementing FromTree trait. This avoids copying all the Pk::Associated type bounds
 /// throughout the codebase.
 macro_rules! impl_block_str {
     ($(;$gen:ident; $gen_con:ident, )* $name: ty,

```

**Review by tcharding — APPROVED** [review]
ACK 8e5b09ba3b47b99d9e44e389a3b81adcb5580aed

**Comment by tcharding** [comment]
Its clean now! Thanks for putting up with my picky reviews :)

**Review by apoelstra — APPROVED** [review]
ACK 8e5b09ba3b47b99d9e44e389a3b81adcb5580aed

**Pull request merged**